libslide/src/partial_evaluator/flatten.rs

```diff
@@ -155,6 +172,318 @@ fn flatten_add_or_sub(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_subtract: bool) ->
+///      mul        div         registry       const
+/// --------------||----
+/// | x | (/ 1 y) || 1 |      {x: 1, y: -1}     2/5
+/// --------------||----
+/// ```
+///
+/// Two steps this time:
+///
+/// ```text
+///  mul    div           registry       const
+/// ----||--------
+/// | 1 || 1 | y |      {x: 2, y: -1}     2/5
+/// ----||--------
+/// ```
+///
+/// Three steps this time:
+///
+/// ```text
+/// m  d        registry       const
+/// -||-
+/// ||||      {x: 2, y: -2}     2/5
+/// -||-
+/// ```
+///
+/// And now, all that needs to be done is to construct the flattened expression `2/5 * x^2 / y^-2`.
+fn flatten_mul_or_div(o_lhs: &Rc<Expr>, o_rhs: &Rc<Expr>, is_div: bool) -> Rc<Expr> {
+    let lhs = flatten_expr(o_lhs);
+    let rhs = flatten_expr(o_rhs);
+
+    let mut coeff = 1.;
+    // Term -> # of times it is multiplied. Negative values are equivalent to division.
+    let mut terms = BTreeMap::<&Rc<Expr>, f64>::new();
+
+    let mut args = VecDeque::with_capacity(2);
+    let base_args = [lhs, rhs];
+    args.extend(base_args.iter());
+
+    // If this is not a division, the first two args are both on the mul side.
+    let mut args_before_div = if is_div { 1 } else { 2 };
+
+    while let Some(arg) = args.pop_front() {
+        let div_side = args_before_div <= 0;
+        args_before_div -= 1;
+
+        let arg = unwrap_expr(arg);
+
+        match arg.as_ref() {
+            Expr::Const(konst) => {
+                if div_side {
+                    coeff /= konst;
+                } else {
+                    coeff *= konst;
+                }
+            }
+            Expr::BinaryExpr(BinaryExpr { op, lhs, rhs })
+                if op == &BinaryOperator::Mult || op == &BinaryOperator::Div =>
+            {
+                if div_side {
+                    if op == &BinaryOperator::Mult {
+                        // 1 / (2 * 3) -> 1 / 2 / 3; add both operands to the div side.
+                        args.push_back(lhs);
+                        args.push_back(rhs);
+                    } else {
+                        // 1 / (2 / 3) -> 3 / 2; here we multiply by the reciprocal, so 3 goes on
+                        // the mul side and 2 goes on the div side.
+                        args.push_front(rhs);
+                        args_before_div = 1;
+                        args.push_back(lhs);
+                    }
+                } else {
+                    // mul side
+                    if op == &BinaryOperator::Mult {
+                        // 1 * (2 * 3) -> 1 * 2 * 3
+                        args.push_front(lhs);
+                        args.push_front(rhs);
+                        args_before_div += 2;
+                    } else {
+                        // 1 * (2 / 3) -> 1 * 2 / 3; add 2 to the mul side and 3 to the div side.
+                        args.push_front(lhs);
+                        args_before_div += 1;
+                        args.push_back(rhs);
+                    }
```

Comment:
this is clean.

Reply:
100% present because of the giant doc block I wrote, it would have been way worse otherwise